src/os/bluestore/BlueStore.cc

```diff
@@ -11084,6 +11084,9 @@ int BlueStore::_upgrade_super()
 	t->rmkey(PREFIX_SUPER, "min_min_alloc_size");
       }
       ondisk_format = 2;
+
+      // make sure upgrade is performed within a single transaction
+      // hence preparing new super record here other than at the end
       _prepare_ondisk_format_super(t);
```

Comment:
Well now I'm confused because AFAIK we just can't ever invoke _prepare_ondisk_format_super() until ondisk_format==4 (for now, the latest version anyway).

So either we have to fix whatever that requires, or else we have to do all the disk format changes in sequence and only invoke _prepare_ondisk_format_super() at the end.

I don't really know which pieces of what are important here so I can't help with those invariants unfortunately.

Reply:
Not sure I understand what's wrong with this version except being a bit sub-optimal. But here is another turn.